api/odilo.py

```diff
@@ -0,0 +1,495 @@
+                        raise CannotFulfill(error_id + ': ' + error_description)
+
+                return content_link, content, content_type
+
+        raise CannotFulfill("Cannot obtain a download link for patron[%r], record_id[%s], format_type[%s].", patron,
+                            record_id, format_type)
+
+    def get_patron_checkouts(self, patron, pin):
+        data = self.patron_request(patron, pin, self.PATRON_CHECKOUTS_ENDPOINT.format(patronId=patron)).json()
+        self.raise_exception_on_error(data)
+        return data
+
+    def get_patron_holds(self, patron, pin):
+        data = self.patron_request(patron, pin, self.PATRON_HOLDS_ENDPOINT.format(patronId=patron)).json()
+        self.raise_exception_on_error(data)
+        return data
+
+    def patron_activity(self, patron, pin):
+        odilo_checkouts = self.get_patron_checkouts(patron, pin)
+        odilo_holds = self.get_patron_holds(patron, pin)
+
+        loans_info = []
+        holds_info = []
+
+        collection = self.collection
+
+        for checkout in odilo_checkouts:
+            loan_info = self.loan_info_from_odilo_checkout(collection, checkout)
+            loans_info.append(loan_info)
+
+        for hold in odilo_holds:
+            hold_info = self.hold_from_odilo_hold(collection, hold)
+            holds_info.append(hold_info)
+
+        return loans_info + holds_info
+
+    def hold_from_odilo_hold(self, collection, hold):
+        start = self.extract_date(hold, 'startTime')
+        # end_date: The estimated date the title will be available for the patron to borrow.
+        end = self.extract_date(hold, 'notifiedTime')
+        position = hold.get('holdQueuePosition')
+
+        if position is not None:
+            position = int(position)
+
+        # Patron already notified to borrow the title
+        if 'informed' == hold['status']:
+            position = 0
+
+        return HoldInfo(
+            collection,
+            DataSource.ODILO,
+            Identifier.ODILO_ID,
+            hold['id'],
+            start_date=start,
+            end_date=end,
+            hold_position=position
+        )
+
+    def place_hold(self, patron, pin, licensepool, notification_email_address):
+        """Place a book on hold.
+
+        :return: A HoldInfo object
+        """
+
+        record_id = licensepool.identifier.identifier
+
+        # Data just as 'x-www-form-urlencoded', no JSON
+        payload = dict(patronId=patron)
+
+        response = self.patron_request(
+            patron, pin, self.PLACE_HOLD_ENDPOINT.format(recordId=record_id),
+            extra_headers={'Content-Type': 'application/x-www-form-urlencoded'},
+            data=payload)
+
+        data = response.json()
+        if response.status_code == 200:
+            return self.hold_from_odilo_hold(licensepool.collection, data)
+
+        error_code, error_description = self.raise_exception_on_error(data)
+        raise CannotHold(error_code + ': ' + error_description)
```

Comment:
Your use of this pattern makes me think it might be useful to pass a default_exception_class into raise_exception_on_error, and change that method to raise an instance of that class if the error message doesn't result in a more specific exception.

Reply:
Good idea, I've added that 'default_exception_class', thank you.